.cursorrules

```diff
@@ -0,0 +1,489 @@
+```cairo
+    tempvar inner_error = new BytesStruct(cast(0, felt*), 0);
+    let err = ErrorType(inner_error);
+```
+
+### Implicit Arguments Pattern
+
+When implementing mutable data structures:
+1. Use implicit arguments for the structure being modified
+2. Return only the error type for modification operations
+3. Return both the value and error for query operations
+
+Example:
+```cairo
+# Modification operation
+func push{stack: Stack}(value: U256) -> StackOverflowError {
+    # Only return error
+}
+
+# Query operation
+func pop{stack: Stack}() -> (U256, StackUnderflowError) {
+    # Return both value and error
+}
+```
+
+### Test Structure Pattern
+
+Tests should:
+1. Be organized in classes (e.g. `TestStack`)
+2. Use the `cairo_run` fixture to run Cairo functions
+3. Compare Python and Cairo implementations for equivalence
+4. Use hypothesis for property-based testing
+5. Test both success and error cases
+6. Have the same path in the tests folder as the source file (e.g. `ethereum/crypto/hash.cairo` -> `tests/ethereum/crypto/test_hash.py`)
+
+To run tests, use pytest with `uv`: `uv run pytest -k <test_name>`.
+
+Example:
+```python
+class TestFeature:
+    @given(...)
+    def test_operation(self, cairo_run, ...):
+        # Test success case
+        result_cairo = cairo_run("operation", ...)
+        result_py = operation(...)
+        assert result_cairo == result_py
+
+        # Test error case
+        with pytest.raises(ErrorType):
+            cairo_run("operation", ...)
+```
+
+### Custom Types
+
+When implementing custom types:
+1. Create a Python dataclass
+2. Implement `gen_arg` method for Cairo memory layout
+3. Add serialization logic in `serde.py`
+4. Register all related types in `_cairo_struct_to_python_type`
+5. Add hypothesis strategies in `strategies.py`
+
+Example:
+```python
+@dataclass
+class CustomList(List[T]):
+    def gen_arg(self, dict_manager, segments):
+        # Convert to Cairo memory layout
+        pass
+
+# Register in args_gen.py
+_cairo_struct_to_python_type = {
+    ("path", "to", "type"): CustomList,
+    ("path", "to", "exceptions"): CustomError,
+}
+
+# Add strategy in strategies.py
+custom_list = st.lists(base_strategy)
+```
+
+### Type Wrapping Pattern
+
+Complex types use pointer-based structures for consistent memory allocation. Two
+main variants exist:
+
+1. Simple wrapper (fixed-size):
+
+```cairo
+struct Bytes0 {
+    value: felt,
+}
+```
+
+2. Complex wrapper (variable-size):
+
+```cairo
+struct Bytes {
+    value: BytesStruct*,
+}
+
+struct BytesStruct {
+    data: felt*,
+    len: felt,
+}
+```
+
+### Optional Values Pattern
+
+Null values implementation uses pointer semantics:
+
+1. Simple types (direct value):
+
+```cairo
+struct U64 {
+    value: felt
+}
+```
+
+A `null` U64 is represented by the outer U64\* with a value of 0.
+
+2. Complex types (pointer-based):
+
+```cairo
+struct U256 {
+    value: U256Struct*,  // null when pointer is 0
```

Comment:
Not sure how it will be interpreted by cursor: a U256 value is _considered_ null if the U256Struct* is 0. To understand this comment you need to put it in context.

```suggestion
    value: U256Struct*,  // [value is] considered null when pointer is 0
```

Maybe I am nitpicking